File: 28. Task scheduling with priorities using pq.cpp
/*
 Task: Simple task scheduling with priorities using priority_queue
*/

#include <bits/stdc++.h>
using namespace std;

int main()
{
    priority_queue<pair<int, string>> pq;

    pq.push({10, "A"});
    pq.push({5, "B"});
    pq.push({20, "C"});

    while (!pq.empty())
    {
        auto p = pq.top();
        pq.pop();
        cout << p.second << " with priority " << p.first << "\n";
    }

    return 0;
}

// Time Complexity: O(n log n)
// Space Complexity: O(n)

File: 28. Task scheduling with priorities using pq.py
"""
Task: Task scheduling with priorities using heapq
Time Complexity: O(n log n)
Space Complexity: O(n)
"""

import heapq


def schedule(tasks):
    # tasks: list of (priority, name)
    pq = []
    for p,n in tasks:
        heapq.heappush(pq, (-p, n))
    while pq:
        p,n = heapq.heappop(pq)
        print(n, 'with priority', -p)

if __name__=="__main__":
    schedule([(10,'A'), (5,'B'), (20,'C')])

# Task: Task scheduling with priorities using pq
